src/hashfunctions.c

```diff
@@ -193,7 +193,7 @@ Int BasicRecursiveHashForPRec(Obj obj)
         UInt rechash = BasicRecursiveHash(GET_ELM_PREC(obj, i));
 
         // Use +, because record may be out of order
-        current_hash += HashCombine2(hashrecname, rechash);
+        current_hash += ShuffleUInt(HashCombine2(hashrecname, rechash));
```

Comment:
Wait, why is this needed?

Uh-oh, is it because we hash (small) integers via identity??? It seems like it, looking at `DATA_HASH_FUNC_FOR_INT`. That's a terrible idea (might very well be my fault, too *sigh*), for many reasons, including the need for this workaround here.

Perhaps we still want this here, to workaround other "bad" hash functions; but then again, using a bad hash function has many ripple effects, I am not sure we can deal with them all in such a way... ah well *sigh*

Reply:
I've submitted PR #148 to deal with this. But it is not enough here because the function `HashCombine2` (and also `HashCombine3`) is linear, which arguably is also problematic...

All in all, I must say I like the Julia design, were all hash functions take an UInt "seed" as a second argument, and are supposed to "mix" this into the hash; in our context that would mean anything using `HASHKEY_MEM_NC` could just use this given seed, and "only" the remaining leaves, such as the hash function for integers, would have to do something special about it. It also makes it very easy to automatically get hashes that depend on order.

Reply:
Yes, actually we probably want a better `HashCombine`, and then we wouldn't need the extra shuffle here.